llama.cpp

```diff
@@ -5820,6 +5826,125 @@ struct llm_tokenizer_bpe {
         return bpe_encoded_words;
     }
 
+    std::vector<std::string> bpe_deepseek_coder_preprocess(const std::string & text) {
+        std::vector<std::string> bpe_words;
+        std::vector<std::string> bpe_encoded_words;
+
+        std::string token = "";
+        
+        bool collecting_letter = false;
+        bool collecting_special = false;
+        bool collecting_punc = false;
+        bool collecting_cjk = false;
+        bool collecting = false;
+
+        std::vector<std::string> text_utf;
+        text_utf.reserve(text.size());
+        bpe_words.reserve(text.size());
+        bpe_encoded_words.reserve(text.size());
+
+        auto cps = codepoints_from_utf8(text);
+        for (size_t i = 0; i < cps.size(); ++i)
+            text_utf.emplace_back(codepoint_to_utf8(cps[i]));
+
+        for (int i = 0; i < (int)text_utf.size(); i++) {
+            const std::string & utf_char = text_utf[i];
+            bool split_condition = false;
+            int bytes_remain = text_utf.size() - i;
+            // std::cout<<utf_char<<std::endl;
+            // forward backward lookups
+            const std::string & utf_char_next = (i + 1 < (int)text_utf.size()) ? text_utf[i + 1] : "";
+            const std::string & utf_char_next_next = (i + 2 < (int)text_utf.size()) ? text_utf[i + 2] : "";
```

Comment:
I don't think C++ allows you to bind a reference to a temporary std::string (converted from "") like this.